src/ripple/core/impl/Stoppable.cpp

```diff
@@ -177,22 +177,16 @@ RootStoppable::isStopping() const
     return m_calledStop;
 }
 
-void
-RootStoppable::prepare()
-{
-    if (m_prepared.exchange(true) == false)
-        prepareRecursive();
-}
-
 void
 RootStoppable::start()
 {
     // Courtesy call to prepare.
-    if (m_prepared.exchange(true) == false)
-        prepareRecursive();
+    if (m_prepared.exchange(true))
+        return;
 
-    if (m_started.exchange(true) == false)
-        startRecursive();
+    prepareRecursive();
+    m_started.exchange(true);
```

Comment:
This doesn't need to be behind a second conditional because it is guarded by the one above. In the old code, one could imagine a scenario where one thread calls `start()`, which enters `prepareRecursive()`, and while it's in there, another thread calls `start()`, which skips the call to `prepareRecursive()` and dives into `startRecursive()` before `prepareRecursive()` has finished in the other thread, potentially violating the assumptions of `Stoppable`s on the call sequence of their `onPrepare` and `onStart` callbacks. That doesn't happen in the code as it was written, but this change makes it impossible.